### ISSUE TITLE
Random noise outputted when sampling after training

### ISSUE THREAD

**Issue by pacocp** [lucidrains/imagen-pytorch#25]
Hi!

First of all, thank you very much for the cool implementation, there are not many diffusion model implementations out there.

I have been trying the model with a biomedical dataset without conditioning on anything. However, while the model is appropriately training (both the unet losses go down) and if I visualize the training predictions the model is able to predict x_start correctly, with the sampling method the model only outputs random noise. 
I have been debugging the code and in the p_sample_loop function, the model is returning random noise images all the time, even though the p_sample function seems to work perfectly.

Has anyone encountered this issue?

Thanks!

**Comment by lucidrains** [comment]
@pacocp hi Francisco! thanks for trying out the repository, and on a biological application to boot! i'll get this all squared away by week's end. It has only been 1 week so far since the release of Imagen and I'm working as fast as I can :) typically in my repositories, if you see a `wip` flag near the title, it means i am not confident yet

**Comment by lucidrains** [comment]
@pacocp but glad to hear the loss is going down, that's a better sign than not going down (or going up, for that matter) lol

**Comment by pacocp** [comment]
@lucidrains sure, and thank you very much for the quick implementation! I just wanted to make sure that I wasn't doing anything weird, I have been tinkering with the code a bit. Thank you again!

**Comment by lucidrains** [comment]
> @lucidrains sure, and thank you very much for the quick implementation! I just wanted to make sure that I wasn't doing anything weird, I have been tinkering with the code a bit. Thank you again!

so if you are using the `ImagenTrainer`, it may default to keeping track of the exponential moving averaged neural nets, and only after a sufficient number of steps will it start updating the EMA versions. you can always try turning off the EMA and see if perhaps the bug lies around there

**Comment by pacocp** [comment]
> > @lucidrains sure, and thank you very much for the quick implementation! I just wanted to make sure that I wasn't doing anything weird, I have been tinkering with the code a bit. Thank you again!
> 
> so if you are using the `ImagenTrainer`, it may default to keeping track of the exponential moving averaged neural nets, and only after a sufficient number of steps will it start updating the EMA versions. you can always try turning off the EMA and see if perhaps the bug lies around there

Sure! I am going to try that, thanks for the hint! I will let you know if that fix the bug

**Comment by pacocp** [comment]
@lucidrains  Just an update, setting ```set_ema=False``` fixed the sampling. In a small experiment with a few hundred images, it is sampling something (mainly noise but I can see the tissue structure).

Thanks again!
![Captura de Pantalla 2022-05-31 a las 11 30 37](https://user-images.githubusercontent.com/7196966/171259412-6d7142d9-9c51-4a07-bf73-0e07d294af29.png)

**Comment by lucidrains** [comment]
@pacocp very nice! machine imagined histology slides seems ripe to disrupt medical education :)

**Comment by lucidrains** [comment]
@pacocp you can also try lowering the `ema_beta` value to something like 0.99, and you'll still have the exponential moving average, but it'll update much quicker

**Comment by pacocp** [comment]
@lucidrains Yes, I was going to try that now! Thanks!
